gtsfm/data_association/point3d_initializer.py

```diff
@@ -58,98 +58,123 @@ class Point3dInitializer(NamedTuple):
+            i2, uv2 = track_2d.measurements[k2]
+
+            # check for unestimated cameras
+            if self.track_camera_dict.get(i1) is None or self.track_camera_dict.get(i2) is None:
+                logger.warning("Unestimated cameras found at indices {} or {}. Skipping them.".format(i1, i2))
+                continue
+
+            camera_estimates = CameraSetCal3Bundler()
+
+            camera_estimates.append(self.track_camera_dict.get(i1))
+            camera_estimates.append(self.track_camera_dict.get(i2))
+
+            img_measurements = Point2Vector()
+            img_measurements.append(uv1)
+            img_measurements.append(uv2)
+
+            # triangulate point for track
+            try:
+                triangulated_pt = gtsam.triangulatePoint3(
+                    camera_estimates,
+                    img_measurements,
+                    rank_tol=SVD_DLT_RANK_TOL,
+                    optimize=True,
+                )
+            except RuntimeError:
+                # TODO: handle cheirality exception properly?
+                logger.info(
+                    "Cheirality exception from GTSAM's triangulatePoint3() likely due to outlier, skipping track"
+                )
+                continue
+
+            errors = self.compute_track_reprojection_errors(track_2d.measurements, triangulated_pt)
+
+            # The best solution should correspond to the one with most inliers
+            # If the inlier number are the same, check the average error of inliers
+            is_inlier = errors < self.reproj_error_thresh
+
+            # tally the number of votes
+            inlier_errors = errors[is_inlier]
+
+            if inlier_errors.size > 0:
+                # only tally error over the inlier measurements
+                avg_error = inlier_errors.mean()
+                num_votes = is_inlier.astype(int).sum()
+
+                if (num_votes > best_num_votes) or (num_votes == best_num_votes and avg_error < best_error):
+                    best_num_votes = num_votes
+                    best_error = avg_error
+                    best_inliers = is_inlier
+
+
+        return best_inliers
+
+
+    def triangulate(self, track_2d: SfmTrack2d) -> Tuple[Optional[SfmTrack], Optional[float], bool]:
         """Triangulates 3D point according to the configured triangulation mode.
 
         Args:
             track: feature track from which measurements are to be extracted
 
         Returns:
             track with inlier measurements and 3D landmark. None returned if triangulation fails or has high error.
+            avg_track_reproj_error: reprojection error of 3d triangulated point to each image plane
+                Note: this may be "None" if the 3d point could not be triangulated successfully
+                due to a cheirality exception or insufficient number of RANSAC inlier measurements
+            is_cheirality_failure: boolean representing whether the selected 2d measurements lead
```

Comment:
nit: this is technically not needed, but I guess it helps avoid convoluted logic to figure out chierality exception?

Reply:
Yeah, we have two types of cheirality exceptions to log: (1) those in RANSAC when selecting inliers, and (2) those when doing final triangulation.

I'm only logging those from part (2) here. To aggregate stats over all tracks, we have to do it in `DataAssociation`, instead of in the `Point3dInitializer`.

I'm not sure if those from (1) should be logged, or where it should be logged.